File: backend/Testing/Browser/Page/AuthenticatedBasePage.cs
using Microsoft.Playwright;
using Testing.Browser.Component;

namespace Testing.Browser.Page;

public class AuthenticatedBasePage<T> : BasePage<T> where T : AuthenticatedBasePage<T>
{
    public EmailVerificationAlert EmailVerificationAlert { get; }
    public AuthenticatedBasePage(IPage page, string url, ILocator testLocator)
    : base(page, url, new[] { page.Locator("button .i-mdi-account-circle"), testLocator })
    {
        EmailVerificationAlert = new EmailVerificationAlert(Page);
    }

    public async Task GoHome()
    {
        await Page.Locator(".breadcrumbs").GetByRole(AriaRole.Link, new() { Name = "Home" }).ClickAsync();
    }
}

File: frontend/tests/pages/authenticatedBasePage.ts
import type {Locator, Page} from '@playwright/test';

import {AuthenticatedDrawer} from '../components/authenticatedDrawer';
import {BasePage} from './basePage';
import {EmailVerificationAlert} from '../components/emailVerificationAlert';

export class AuthenticatedBasePage extends BasePage {
  readonly emailVerificationAlert: EmailVerificationAlert;

  private drawerToggle: Locator;

  constructor(page: Page, locator: Locator | Locator[], url?: string) {
    const drawerToggle = page.locator('label .i-mdi-account-circle');
    if (Array.isArray(locator)) {
      locator = [drawerToggle, ...locator];
    } else {
      locator = [drawerToggle, locator];
    }
    super(page, locator, url);
    this.drawerToggle = drawerToggle;
    this.emailVerificationAlert = new EmailVerificationAlert(page);
  }

  clickHome(): Promise<void> {
    return this.page.locator('.breadcrumbs').getByRole('link', {name: 'Home'}).click();
  }

  async openDrawer(): Promise<AuthenticatedDrawer> {
    await this.drawerToggle.click();
    return new AuthenticatedDrawer(this.page).waitFor();
  }
}
